policy-controller/k8s/index/src/index.rs

```diff
@@ -0,0 +1,583 @@
+                    .port_servers
+                    .iter()
+                    .filter_map(|(port, ps)| {
+                        if ps.name.as_ref() == Some(&server.name) {
+                            Some(port)
+                        } else {
+                            None
+                        }
+                    })
+                    .copied()
+                    .collect::<Vec<_>>();
+                for port in server_ports.into_iter() {
+                    pod.set_default_server(port, &self.cluster_info);
+                }
+            }
+        }
+    }
+
+    /// Deletes a Server from the index, reverting all pods that use it to use their default server.
+    ///
+    /// Returns true if the Server was deleted and false if it did not exist.
+    pub(crate) fn delete_server(&mut self, name: &str) {
+        if self.servers.remove(name).is_none() {
+            return;
+        }
+
+        for pod in self.pods.values_mut() {
+            for (port, ps) in pod.port_servers.iter_mut() {
+                if ps.name.as_deref() == Some(name) {
+                    ps.name = None;
+                    let server = default_inbound_server(*port, &pod.settings, &*self.cluster_info);
+                    ps.tx.send(server).expect("receiver is held by the index");
+                }
+            }
+        }
+    }
+
+    /// Adds or updates a ServerAuthorization.
+    ///
+    /// Returns true if the ServerAuthorization was updated and false if it already existed and was
+    /// unchanged.
+    pub(crate) fn apply_server_authorization(
+        &mut self,
+        name: impl ToString,
+        server_selector: ServerSelector,
+        authz: ClientAuthorization,
+    ) {
+        let server_authz = ServerAuthorization {
+            authz,
+            server_selector,
+        };
+        let server_authz = match self.server_authorizations.entry(name.to_string()) {
+            Entry::Vacant(entry) => entry.insert(Arc::new(server_authz)).clone(),
+            Entry::Occupied(entry) => {
+                let saz = entry.into_mut();
+                if **saz == server_authz {
+                    return;
+                }
+                *saz = Arc::new(server_authz);
+                saz.clone()
+            }
+        };
+
+        for (srvname, server) in self.servers.iter() {
+            if server_authz.server_selector.selects(server) {
+                let update = mk_inbound_server(
+                    server,
+                    mk_client_authzs(server, &self.server_authorizations),
+                );
+                for pod in self.pods.values_mut() {
+                    if server.pod_selector.matches(&pod.labels) {
+                        for port in pod.select_ports(&server.port_ref).into_iter() {
+                            pod.update_server(port, srvname.to_string(), update.clone());
+                        }
+                    }
+                }
+            }
+        }
+    }
```

Comment:
If the server_selector changed, we also need to update pod servers that are no longer selected...